proposals/0000-splice-imports.rst

```diff
@@ -0,0 +1,319 @@
+
+The new language extension ``ExplicitSpliceImports`` adds a
+new import modifier to the import syntax. An import is marked as a "splice"
+import when it is prefixed with ``splice``::
+
+  {-# LANGUAGE ExplicitSpliceImports #-}
+  {-# LANGUAGE TemplateHaskell #-}
+  module Main where
+
+  -- (1)
+  import splice B
+
+  -- (2)
+  import A
+
+
+The splice modifier indicates to the compiler that module B is only used at compile time
+and hence the imports can **only** be used inside top-level splices (1). When the extension is enabled,
+imports without the splice modifier are only available at runtime and therefore not available to be used in top-level splices (2).
+In this example, identifiers from ``B`` can **only** be used in top-level splices
+and identifiers from ``A`` can be used everywhere, apart from in top-level splices.
+
+To make some of the initial motivation explicit:
+
+1. Now when compiling module ``Main``, despite the fact ``TemplateHaskell`` is enabled,
+   we know that only identifiers from module ``B`` will be used in top-level splices so
+   only ``B`` (and its dependencies) needs to compiled to object code before starting to compile ``Main``.
+2. When cross-compiling, only ``A`` needs to be built for the target and ``B``
+   only for the host as it is only used at build-time.
+
+If you require scenario (3) then two imports declarations can be used::
+
+  -- (3)
+  import C
+  import splice C
+
+The syntax for imports is changed in the follow way::
+
+  importdecl :: { LImportDecl GhcPs }
+     : 'import' maybe_src maybe_safe optsplice optqualified maybe_pkg modid optqualified maybeas maybeimpspec
+
+
+The ``splice`` keyword appears before the ``qualified`` keyword but after ``SOURCE``
+and ``SAFE`` pragmas.
+
+
+Intuitive Specification of ``splice``
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Identifiers arising from splice imports can only be used at negative levels, ie, unquoted in a top-level splice::
+
+  -- Accepted, because B is a splice import and B.qux is used at level -1
+  foo = $(B.qux)
+
+  -- Rejected, because B is a splice import and B.qux is used at level 0
+  foo' =  B.qux
+
+
+But identifiers from normal imports are rejected::
+
+  -- Rejected, as A is not a splice import and used at level -1
+  baz = $(A.zee)
+
+An identifier can appear inside a top-level splice, if it is at a non-negative
+level. For example, the following is legal::
+
+  foo = $(B.qid [| A.zee |] )
+
+Because ``A.zee`` is used at level 0 it doesn't need to be imported using a splice import.
+
+Level Specification of ``splice``
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+* Ordinary imports introduce variables at all non-negative levels (>= 0)
+* Splice imports introduce variables at all negative levels. (< 0)
+
+Ambiguity of instances and variables
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Resolution of scopes (often called "renaming") is blind to whether or not an
```

Comment:
Actually, now I’m somewhat confused because there seems to be a contradiction between these statements in the proposal:

1. “Therefore the `splice` modifier affects name resolution depending on whether we are inside a top-level splice or not.”
2. “Resolution of scopes (often called "renaming") is blind to whether or not an identifier was imported with `splice`.”

Perhaps the design was changed during discussion (which I have not followed closely) but the text was not updated?

Reply:
This should have been clarified. Name resolution is not affected by splice imports (as you suggested above), typechecking is affected, and errors are produced if an import is used at the wrong level.